server/mqtt.go

```diff
@@ -2535,22 +2592,30 @@ func (as *mqttAccountSessionManager) serializeRetainedMsgsForSub(rms map[string]
 // match the given subscription's `subject` (which could have wildcards).
 //
 // Account session manager NOT lock held on entry.
-func (as *mqttAccountSessionManager) loadRetainedMessagesForSubject(rms map[string]*mqttRetainedMsg, topSubject []byte, log Logger) {
+func (as *mqttAccountSessionManager) loadRetainedMessagesForSubject(topSubject []byte, log Logger) int {
 	as.mu.RLock()
 	if len(as.retmsgs) == 0 {
 		as.mu.RUnlock()
-		return
+		return 0
 	}
 	result := as.sl.ReverseMatch(string(topSubject))
 	as.mu.RUnlock()
 
 	if len(result.psubs) == 0 {
-		return
+		return 0
 	}
+	numToDeliver := 0
 	for _, sub := range result.psubs {
 		subject := string(sub.subject)
-		if rms[subject] != nil {
-			continue // already loaded
+		ref := as.retmsgs[subject]
+		if ref == nil {
+			// If we don't have anything for the subject - skip, but this should
+			// not happen.
+			continue
+		}
+		if ref.loaded != nil && ref.loaded.expiresFromCache.After(time.Now()) {
```

Comment:
That would be wrong too and be racy since this is not protected by the `as`'s lock. Also, if we want to use the `ref.loaded` we should either bump the expiration or turnoff expiration for that message until we are done with it.